src/transformers/modeling_blenderbot.py

```diff
@@ -0,0 +1,582 @@
+import math
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from torch.nn import LayerNorm
+from transformers.modeling_utils import PreTrainedModel
+from configuration_blenderbot import BlenderbotConfig
+import numpy as np
+import warnings
+
+BLENDERBOT_PRETRAINED_MODEL_ARCHIVE_MAP = {"blenderbot": "https://cdn.huggingface.co/"}
+
+
+class BlenderbotMultiHeadAttention(nn.Module):
+    """
+    multiheadattention from vaswani paper
+    """
+
+    def __init__(self, n_heads, dim, dropout=0.0):
+        """
+
+        :param n_heads: number of attention heads
+        :param dim: dimension
+        :param dropout: attention dropout
+        """
+        super(BlenderbotMultiHeadAttention, self).__init__()
+        self.n_heads = n_heads
+        self.dim = dim
+        self.att_dropout = nn.Dropout(p=dropout)
+        self.k_lin = nn.Linear(dim, dim)
+        self.q_lin = nn.Linear(dim, dim)
+        self.v_lin = nn.Linear(dim, dim)
+        self.out_lin = nn.Linear(dim, dim)
+
+        # weight initialization will be done using the pretrained model
+
+    def forward(self, query, key=None, value=None, mask=None, incremental_state=None, static_key_val=False):
+        """
+
+        :param query: attention query: torch.tensor
+        :param key: attention key: torch.tensor, default to None
+        :param value: attention value: torch.tensor, default to None
+        :param mask: allow attention or not: torch.tensor, default to None
+        :param incremental_state: dictionary with values representing the previous states of the key, value, and mask
```

Comment:
Awesome that you add the cached key, value states here :-) Maybe rename it to `past_key_value_states` as it's done in other models (Reformer, T5, GPT2)